openquake/hazardlib/gsim/base.py

```diff
@@ -212,6 +221,123 @@ def _get_poes(mean_std, loglevels, truncation_level, squeeze=False):
     return _truncnorm_sf(truncation_level, out)
 
 
+def _get_poes_site(mean_std, loglevels, truncation_level, ampfun,
+                   mag, sitecode, rrup, squeeze=False):
+    """
+    NOTE: this works for a single site. We also assume that the loglevels
+    are the same for all the IMTs
+
+    :param mean_std:
+        See :function:`openquake.hazardlib.gsim.base.get_poes`
+    :param loglevels:
+        Intensity measure level per intensity measure type. See
+        :function:`openquake.hazardlib.gsim.base.get_poes`
+    :param truncation_level:
+        The level of truncation of the normal distribution of ground-motion
+        on rock
+    :param ampl:
+        Site amplification function instance of
+        :class:openquake.hazardlib.site_amplification.AmpFunction
+    :param mag:
+        The magnitude of the earthquake
+    :param rrup:
+        The rrup distances
+    :param squeeze:
+        A boolean. Should be True when ...
+    """
+
+    # Mean and std of ground motion for the IMTs considered in this analysis
+    # N  - Number of sites
+    # LS - Number of intensity measure levels
+    # G  - Number of GMMs
+    mean, stddev = mean_std  # shape (N, M, G)
+    N, LS, G = len(mean), len(loglevels.array), mean.shape[-1]
+
+    # List of the IMTs
+    imts = [s for s in loglevels]
+
+    # This is the array where we store the output results i.e. poes on soil
+    out_s = numpy.zeros((N, LS) if squeeze else (N, LS, G))
+
+    # Get the values of ground-motion used to compute the probability of
+    # exceedance on soil. NOTE we assume they are the same for all the
+    # IMTs considered.
+    soillevels = loglevels[imts[0]]
+
+    # `nsamp` is the number of IMLs per IMT used to compute the hazard on rock
+    # while 'L' is total number of ground-motion values
+    nsamp = 40
+
+    # Compute the probability of exceedance for each in intensity
+    # measure type IMT
+    lvl = 0
+    clm = 0
+    for m, imt in enumerate(loglevels):
+
+        # Here we set automatically the IMLs that will be used to compute
+        # the probability of occurrence of GM on rock within discrete
+        # intervals
+        sigma = ampfun.get_max_sigma()
+        ll = numpy.linspace(min(soillevels)-sigma*4.,
+                            max(soillevels)+sigma*4., num=nsamp)
+
```

Comment:
There lines can go outside the loop, right?

Reply:
I fixed this part since the IMLs can vary depending on the IMT. We can move out just the call to get_max_sigma.